apps/omg_eth/lib/omg_eth/ethereum_height_monitor.ex

```diff
@@ -163,15 +163,16 @@ defmodule OMG.Eth.EthereumHeightMonitor do
     end
   end
 
-  @spec broadcast_on_new_height(module(), non_neg_integer(), non_neg_integer() | :error) :: :ok | {:error, term()}
-  defp broadcast_on_new_height(_event_bus, _previous_height, :error), do: :ok
+  @spec broadcast_on_new_height(module(), non_neg_integer() | :error) :: :ok | {:error, term()}
+  defp broadcast_on_new_height(_event_bus, :error), do: :ok
 
-  defp broadcast_on_new_height(event_bus, previous_height, height) when height > previous_height do
+  # we need to publish every height we fetched so that we can re-examine blocks in case of re-orgs
+  # clients subscribed to this topic need to be aware of that and if a block number repeats,
+  # it needs to re-write logs, for example
+  defp broadcast_on_new_height(event_bus, height) do
```

Comment:
whatever service is consuming this process events, it should solve it's own problems to regards to ethereum reorgs. if we don't publish "the same height", we might hide a possible blockchain re-org event

Reply:
because detecting new blocks is difficult (they're not timed precisely) and we fetch new height every 8seconds instead of ~15, just to not miss a new one, there would be a lot more noise and code added to support that behaviour (I don't see a use case really atm). 
I guess the logic would then need to account that if the previous block is equal or smaller then the new block - this is a potential reorg. if it's indeed smaller  - it's a reorg. but if it's the same *it might be*. now this *might be* would be noisy.
1. because we fetch for new height aggressively
2. because reorgs of height 1 happen quite often

Reply:
Blockchain huh. So are the consumers dealing with all stuff currently? :/

Reply:
there's only one consumer, which is the `OMG.Eth.EthereumHeight` and this process doesn't directly deal with reorgs, it doesn't care... the only api it has is `get/0` which returns the current ethereum height. 

ethereum event listeners have a margin of around 10 blocks, so they're on the safe side of a reorg happening

Reply:
block getter on the other hand ... @pdobacz can you 💡 ? I think it's safe because of the safety net on the childchain side

Reply:
hm, so block getter doesn't wait for finality margin. The Block Getter is driven by the _child chain block_ number, as it appears to be when the root chain contract is called.

The Block Getter should withstand reorgs and unexpected back-offs of the ethereum height reported. What it does it looks for the event logs which "accompany" the submitted blocks (`BlockSubmitted`), just to locate the submissions with respect to the Ethereum height where they appeared.

Reply:
*without looking at the code*
how does Block Getter get the block range it needs to scan for block emitted events? the (from_block, to_block)

Reply:
_without looking at the code too :joy:_

Block Getter, like other Ethereum-syncers mainains its `synced_height`. To cover reorgs it looks for those events in vicinity of that height (+/- ~1000 blocks IIRC), whenever it needs to associate downloaded child blocks with `BlockSubmitted` events/ethereum heights.

OK, I had some doubts and dug deeper, but what I wrote is quite accurate so far:
 - https://github.com/omisego/elixir-omg/blob/master/apps/omg_watcher/lib/omg_watcher/block_getter.ex#L280
 - then https://github.com/omisego/elixir-omg/blob/master/apps/omg_watcher/lib/omg_watcher/block_getter/core.ex#L218
 - `coordinator_height` is the "can sync up to X" figure calculated by the `RootChainCoordinator`
 - among other factors, it will not be more than a fixed amount https://github.com/omisego/elixir-omg/blob/master/apps/omg/lib/omg/root_chain_coordinator/core.ex#L127

The values of the 1000 and 2500 constants where figured out by trial-and-error and educated guesses, there isn't any science behind them.

If I'm not mistaken, the changes here should interfere with this behavior, at least not "critically". At most maybe things (constants) might require tweaking. How do you think?

Reply:
oh nice! now I _get_ it. so one important configuration I wasn't aware of is:
```

* **`block_getter_reorg_margin`** - the margin considered by `OMG.Watcher.BlockGetter` when searching for recent child chain block submission events.
This is driving the process of determining the height and particular event related to the submission of a particular child chain block
```
```
block_getter_reorg_margin: 200
```
Isn't that a lot? Wouldn't 10 blocks suffice? 
> If I'm not mistaken, the changes here should interfere with this behavior, at least not "critically". At most maybe things (constants) might require tweaking. How do you think?

Did you mean shouldn't? The tests I've made the responses from Infura were pretty constant. Since the requests now include all events, the responses can get large. We haven't really tried with a block FULL of all sorts of events. That would be a good reliability test though.

Reply:
> Did you mean shouldn't?

:man_facepalming: fixed

> `block_getter_reorg_margin: 200`

Hard to say, if this is a lot or not. `10` seems a little too small. This is because if it so happens that a `BlockSubmitted` you're expecting to see within this range of blocks (only 10 back) gets reorged to a block 11 blocks behind your `synced_height` (what `BlockGetter` processed till now), it will then appear that there were no submissions in that range and the `BlockGetter` will go looking elsewhere and get confused.

Since we don't want that, the pretty generous 200 is used, which is still modest, compared to the fact that you want (or at least we wanted to) cover 1000s of blocks, when catching up with events which are "sparse" (this in turn makes you avoid doing 1000s of individual RPC calls).

200 makes me feel confident as an isolated block submission jumping this far back is very unlikely. 10 is very unlikely too but makes me feel more uneasy.